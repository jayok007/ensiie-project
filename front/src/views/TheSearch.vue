<template>
  <div>
    <search-bar
      :keywords="keywords"
      v-model="selectedKeywords"
      @search="search"
    ></search-bar>

    <v-container fluid grid-list-xl>
      <v-progress-linear
        v-if="loading"
        color="success"
        height="4"
        indeterminate
      ></v-progress-linear>

      <v-alert
        :value="alert"
        type="info"
      >
        Désolé, mais votre recherche n'a rien donné.
        Essayez avec d'autres mots clés !
      </v-alert>

      <v-layout row wrap>
        <bar
          v-for="{ id, name, address, keywords } in bars"
          :key="id"
          v-bind:id="id"
          :name="name"
          :address="address"
          :keywords="keywords"
        ></bar>
      </v-layout>
    </v-container>
  </div>
</template>

<script>
import Bar from '@/components/Bar'
import SearchBar from '@/components/SearchBar'

export default {
  name: 'TheSearch',

  components: {
    Bar,
    SearchBar
  },

  props: {
    query: {
      type: Object,
      required: true
    }
  },

  data () {
    return {
      keywords: [],
      selectedKeywords: [],
      bars: [],
      loading: true,
      alert: false
    }
  },

  created () {
    this.$api.getKeywords()
      .then(keywords => (this.keywords = keywords))
      .catch(this.$log.error)
  },

  watch: {
    query: {
      immediate: true,
      handler () {
        this.$log.debug(this.query)

        this.selectedKeywords = this.query.q.split(',')
        this.loading = true

        this.$api.getBars(this.query)
          .then(bars => {
            this.bars = bars
            this.alert = false
          })
          .catch(err => {
            this.$log.error(err)

            if (err.response.status === 404) this.alert = true
          })
          .finally(() => (this.loading = false))
      }
    }
  },

  methods: {
    search () {
      if (this.selectedKeywords.join(',') === this.query.q) return
      if (this.selectedKeywords.length === 0) return

      this.bars = []
      this.loading = true
      this.$router.push(`/search?q=${this.selectedKeywords.join(',')}`)
    }
  }
}
</script>
