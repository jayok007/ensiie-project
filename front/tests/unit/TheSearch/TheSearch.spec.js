import { shallowMount } from '@vue/test-utils'
import TheSearch from '@/views/TheSearch'

describe('TheSearch', () => {
  it('should render an empty result', () => {
    const wrapper = shallowMount(TheSearch, {
      propsData: {
        query: { q: 'a,b,c' }
      },
      mocks: {
        $log: {
          debug: jest.fn()
        },
        $api: {
          getBars () {
            return Promise.resolve([])
          }
        },
        $store: {
          dispatch: jest.fn(),
          state: {
            keywords: ['a', 'b', 'c', 'd', 'e']
          }
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a list of bars', async () => {
    const wrapper = shallowMount(TheSearch, {
      propsData: {
        query: { q: 'a,b,c' }
      },
      mocks: {
        $log: { debug: jest.fn() },
        $api: {
          getBars () {
            return Promise.resolve([
              {
                id: 1,
                name: 'Bar 1',
                address: '1 rue',
                keywords: ['b'],
                photoreference: 'ref',
                rating: '4.4'
              },
              {
                id: 2,
                name: 'Bar 2',
                address: '2 rue',
                keywords: ['a'],
                photoreference: 'ref',
                rating: '2.4'
              }
            ])
          }
        },
        $store: {
          dispatch: jest.fn(),
          state: {
            keywords: ['a', 'b', 'c', 'd', 'e']
          }
        }
      }
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
