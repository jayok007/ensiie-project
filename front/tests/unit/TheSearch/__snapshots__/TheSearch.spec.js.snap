// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheSearch should render a list of bars 1`] = `
<div>
  <searchbar-stub keywords="a,b,c,d,e" value="a,b,c"></searchbar-stub>
  <vcontainer-stub tag="div" fluid="" grid-list-xl="">
    <vprogresslinear-stub active="true" buffervalue="100" color="secondary" height="4" indeterminate="true" value="0"></vprogresslinear-stub>
    <valert-stub type="info">
      Désolé, mais votre recherche n'a rien donné.
      Essayez avec d'autres mots clés !
    </valert-stub>
    <vlayout-stub tag="div" row="" wrap="">
      <bar-stub id="1" name="Bar 1" address="1 rue" keywords="b" photoreference="ref" rating="4.4"></bar-stub>
      <bar-stub id="2" name="Bar 2" address="2 rue" keywords="a" photoreference="ref" rating="2.4"></bar-stub>
    </vlayout-stub>
  </vcontainer-stub>
  <vbtn-stub activeclass="v-btn--active" tag="button" type="button" color="secondary">
    Un bar manque à la liste ?
  </vbtn-stub>
</div>
`;

exports[`TheSearch should render an empty result 1`] = `
<div>
  <searchbar-stub keywords="a,b,c,d,e" value="a,b,c"></searchbar-stub>
  <vcontainer-stub tag="div" fluid="" grid-list-xl="">
    <vprogresslinear-stub active="true" buffervalue="100" color="secondary" height="4" indeterminate="true" value="0"></vprogresslinear-stub>
    <valert-stub type="info">
      Désolé, mais votre recherche n'a rien donné.
      Essayez avec d'autres mots clés !
    </valert-stub>
    <vlayout-stub tag="div" row="" wrap=""></vlayout-stub>
  </vcontainer-stub>
  <vbtn-stub activeclass="v-btn--active" tag="button" type="button" color="secondary">
    Un bar manque à la liste ?
  </vbtn-stub>
</div>
`;
